horace_core/sqw/@sqw/sqw_eval.m

```diff
@@ -1,102 +1,348 @@
 %              above then pars = {p, c1, c2, ...}
 %
-% Optional string parameters:
-%   'all'      Requests that the calculated sqw be returned over
+% Keyword Arguments:
+% ------------------
+%   outfile    If present, the outputs will be written to the file of the given
+%              name/path.
+%              If numel(win) > 1, outfile must be omitted or a cell array of
+%              file paths with equal number of elements as win.
+%
+%   all        Requests that the calculated sqw be returned over
 %              the whole of the domain of the input dataset. If not given, then
 %              the function will be returned only at those points of the dataset
 %              that contain data.
-%               Applies only to input with no pixel information - it is ignored if
+%              Applies only to input with no pixel information - it is ignored if
 %              full sqw object.
 %
-%   'average' Requests that the calculated sqw be computed for the
-%              average values of h,k,l of the pixels in a bin, not for each
+%   average    Requests that the calculated sqw be computed for the
+%              average values of h, k, l of the pixels in a bin, not for each
 %              pixel individually. Reduces cost of expensive calculations.
-%              Applies only to the case of sqw object with pixel information 
+%              Applies only to the case of sqw object with pixel information
 %             - it is ignored if dnd type object.
 %
-% Note: all optional string input parameters can be truncated up to minal 
+%   filebacked  If true, the result of the function will be saved to file and
+%               the output will be a file path. If no `outfile` is specified,
+%               a unique path within `tempdir()` will be generated.
+%               Default is false.
+%
+% Note: all optional string input parameters can be truncated up to minimal
 %       difference between them e.g. routine would accept 'al' and
 %       'av', 'ave', 'aver' etc....
 %
 %
 % Output:
 % -------
-%   wout        Output dataset or array of datasets
-
-
-% Check optional argument
-options = {'all','average'};
-[ok,mess,all_bins,ave_pix]=parse_char_options(varargin,options);
-if ~ok
-    error('SQW_EVAL:invalid_argument',mess);
+%   wout        If `filebacked` is false, an sqw object or array of sqw objects.
+%               If `filebacked` is true, a file path or cell array of file paths.
+%               Output argument must be specified if `outfile` not given.
+%
+[sqwfunc, pars, opts] = parse_arguments(win, sqwfunc, pars, varargin{:});
+if isempty(opts.outfile) || opts.filebacked
+    nargoutchk(1, 1);
 end
 
 wout = copy(win);
-if ~iscell(pars), pars={pars}; end  % package parameters as a cell for convenience
+if ~iscell(pars)
+    pars = {pars};  % package parameters as a cell for convenience
+end
+
+for i = 1:numel(win)
 
-for i=1:numel(win)
     if is_sqw_type(win(i))   % determine if sqw or dnd type
```

Comment:
Checking if we should say `has_pixels` - may have to await Nick's new sqw

Reply:
`has_pixels` doesn't exist on `master`, so we'd probably need to wait for Nick.